chainer/functions/activation/sigmoid.py

```diff
@@ -12,51 +12,88 @@
         type_check.expect(in_types[0].dtype.kind == 'f')
 
-    def forward_cpu(self, x):
-        half = x[0].dtype.type(0.5)
-        y = utils.force_array(numpy.tanh(x[0] * half) * half + half)
+    def forward_cpu(self, inputs):
+        x = inputs[0]
+        half = x.dtype.type(0.5)
+        y = utils.force_array(numpy.tanh(x * half) * half + half)
+        self.retain_inputs(())
         self.retain_outputs((0,))
+        self._use_cudnn = False
         return y,
 
     def forward_gpu(self, inputs):
         x = inputs[0]
         if chainer.should_use_cudnn('==always') and x.flags.c_contiguous:
             y = cudnn.activation_forward(x, _mode)
+            self.retain_inputs((0,))
+            self._use_cudnn = True
         else:
             y = cuda.elementwise(
                 'T x', 'T y', 'y = tanh(x * 0.5) * 0.5 + 0.5',
                 'sigmoid_fwd')(x)
             self.retain_inputs(())
+            self._use_cudnn = False
+
         self.retain_outputs((0,))
         return y,
 
-    def backward_cpu(self, x, gy):
-        one = x[0].dtype.type(1)
-        y = self.output_data[0]
-        return utils.force_array(gy[0] * y * (one - y)),
+    def backward(self, indexes, grad_outputs):
+        if self._use_cudnn:
+            x = self.get_retained_inputs()[0].data
+        else:
+            x = None
+        y = self.get_retained_outputs()[0]
+        gy, = grad_outputs
+        return SigmoidGrad((x,)).apply((y, gy))
 
-    def backward_gpu(self, inputs, grads):
-        x = inputs[0]
-        gy = grads[0]
-        y = self.output_data[0]
+
+class SigmoidGrad(function_node.FunctionNode):
+
+    """Logistic sigmoid gradient function."""
+
+    def __init__(self, inputs):
+        super(SigmoidGrad, self).__init__()
+        self.x = inputs[0]
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 2)
+        type_check.expect(in_types[0].dtype.kind == 'f')
+        type_check.expect(in_types[1].dtype.kind == 'f')
+
+    def forward_cpu(self, inputs):
+        self.retain_inputs((0, 1))
+        y, gy = inputs
+        one = y.dtype.type(1)
+        return utils.force_array(gy * y * (one - y)),
+
+    def forward_gpu(self, inputs):
+        self.retain_inputs((0, 1))
+        y, gy = inputs
         if (chainer.should_use_cudnn('==always') and gy.flags.c_contiguous and
-                x is not None and x.flags.c_contiguous):
-            gx = cudnn.activation_backward(x, y, gy, _mode)
+                self.x is not None and self.x.flags.c_contiguous):
+            gx = cudnn.activation_backward(self.x, y, gy, _mode)
         else:
             gx = cuda.elementwise(
                 'T y, T gy', 'T gx',
                 'gx = gy * y * (1 - y)',
                 'sigmoid_bwd')(y, gy)
         return gx,
 
+    def backward(self, indexes, grad_outputs):
+        y, gy = self.get_retained_inputs()
+        g, = grad_outputs
+        one = y.dtype.type(1)
```

Comment:
These `one` and `two` are not needed in `backward` because scalar-Variable arithmetics resolve the type promotion correctly.

Reply:
just use 1 and 2.